railties/lib/rails/engine/configuration.rb

```diff
@@ -44,7 +44,7 @@ def initialize(root = nil)
         @generators = app_generators.dup
         @middleware = Rails::Configuration::MiddlewareStackProxy.new
         @javascript_path = "javascript"
-        @route_set_class = ActionDispatch::Routing::RouteSet
+        @route_set_class = Rails.env.local? ? Rails::Engine::LazyRouteSet : ActionDispatch::Routing::RouteSet
```

Comment:
this is the safest way as when `routes` gets accessed outisde of the initialization process, `@route_set_class` will be properly set

Reply:
The problem is we have specific use-cases (mostly tests) that draw routes early and expect the routes to be there without initialization. In this case, we can't use lazy routes because the initialization process is never fully completed. I think we need to find a different way to approach exceptions app setting.